engine/src/main/java/io/camunda/zeebe/engine/processing/common/ExpressionProcessor.java

```diff
@@ -32,13 +33,45 @@
   private final DirectBuffer resultView = new UnsafeBuffer();
 
   private final ExpressionLanguage expressionLanguage;
-  private final VariableStateEvaluationContext evaluationContext;
+  private final ContextLookup contextLookup;
 
   public ExpressionProcessor(
       final ExpressionLanguage expressionLanguage, final VariablesLookup lookup) {
+
+    this(expressionLanguage, new VariableStateEvaluationContextLookup(lookup));
+  }
+
+  private ExpressionProcessor(
+      final ExpressionLanguage expressionLanguage, final ContextLookup lookup) {
     this.expressionLanguage = expressionLanguage;
+    contextLookup = lookup;
+  }
 
-    evaluationContext = new VariableStateEvaluationContext(lookup);
+  /**
+   * Returns a new {@code ExpressionProcessor} instance. This new instance will use {@code
+   * secondaryContext} for all lookups which it cannot find in its primary evaluation context
+   *
+   * @param secondaryContext fallback evaluation context
+   * @return new instance which uses {@code secondaryContext} as fallback
+   */
+  public ExpressionProcessor withSecondaryContext(final EvaluationContext secondaryContext) {
+    final ContextLookup combinedLookup =
+        scopeKey -> contextLookup.getContext(scopeKey).combine(secondaryContext);
+    return new ExpressionProcessor(expressionLanguage, combinedLookup);
+  }
+
+  /**
+   * Returns a new {@code ExpressionProcessor} instance. This new instance will use {@code
+   * primaryContext} for all lookups. Only if it doesn't find a variable in {@code primaryContext},
+   * it will lookup variables in the evaluation context of {@code this} evaluation processor
+   *
+   * @param primaryContext new top level evaluation context
+   * @return new instance which uses {@code primaryContext} as new top level evaluation context
+   */
+  public ExpressionProcessor withPrimaryContext(final EvaluationContext primaryContext) {
```

Comment:
🔧  This is a bit of a nitpick, but I suggest placing `withPrimaryContext` above the `withSecondaryContext`. The methods then appear in a natural order.